File: clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/internal/StatefulHost.java
package com.algolia.internal;

import com.algolia.config.CallType;
import com.algolia.config.Host;
import com.algolia.utils.DateTimeUtils;
import java.time.OffsetDateTime;
import java.util.Set;

public final class StatefulHost {

  private final Host host;
  private boolean up = true;
  private int retryCount;
  private OffsetDateTime lastUse = DateTimeUtils.nowUTC();

  public StatefulHost(Host host) {
    this.host = host;
  }

  public String getHost() {
    return host.getUrl();
  }

  public String getScheme() {
    return this.host.getScheme();
  }

  public boolean isUp() {
    return up;
  }

  public StatefulHost setUp(boolean up) {
    this.up = up;
    return this;
  }

  public int getRetryCount() {
    return retryCount;
  }

  public StatefulHost resetCount() {
    this.retryCount = 0;
    return this;
  }

  public void incrementRetryCount() {
    this.retryCount++;
  }

  public OffsetDateTime getLastUse() {
    return lastUse;
  }

  public StatefulHost setLastUse(OffsetDateTime lastUse) {
    this.lastUse = lastUse;
    return this;
  }

  public Set<CallType> getAccept() {
    return this.host.getCallTypes();
  }
}

File: clients/algoliasearch-client-scala/src/main/scala/algoliasearch/internal/StatefulHost.scala
package algoliasearch.internal

import algoliasearch.config.{CallType, Host}
import algoliasearch.internal.util.currentDateTime

import java.time.OffsetDateTime
import scala.collection.Set

private[algoliasearch] class StatefulHost(private val host: Host) {
  private var up: Boolean = true
  private var retryCount: Int = 0
  private var lastUse: OffsetDateTime = currentDateTime()

  def getHost: String = host.url

  def getScheme: String = host.scheme

  def isUp: Boolean = up

  def getRetryCount: Int = retryCount

  def incrementRetryCount(): Unit = {
    retryCount += 1
  }

  def getLastUse: OffsetDateTime = lastUse

  def getAccept: Set[CallType] = host.callTypes

  def reset(): Unit = {
    up = true
    lastUse = currentDateTime()
    retryCount = 0
  }

  def hasTimedOut(): Unit = {
    up = true
    lastUse = currentDateTime()
    retryCount += 1
  }

  def hasFailed(): Unit = {
    up = false
    lastUse = currentDateTime()
  }
}

object StatefulHost {
  def apply(host: Host): StatefulHost = new StatefulHost(host)
}
